bundle/edu.gemini.p2checker/src/main/java/edu/gemini/p2checker/checker/P2Checker.java

```diff
@@ -135,16 +125,26 @@ private IRule _getRule(ObservationElements elements) {
      */
     public IP2Problems check(ISPNode node, AgsMagnitude.MagnitudeTable mt) {
         //obs/seq component require us to find the observation first
-        if (node instanceof ISPSeqComponent ||
-                node instanceof ISPObsComponent) {
-            if (node.getParent() instanceof ISPObservation) {
-                return _checkObservation((ISPObservation) node.getParent(), mt);
-            } else {
-                return check(node.getParent(), mt); //recursively attempt to find the Observation
-            }
+        if (node instanceof ISPSeqComponent || node instanceof ISPObsComponent) {
+            return check(node.getContextObservation(), mt);
+
             //observations can be checked immediately
         } else if (node instanceof ISPObservation) {
-            return _checkObservation((ISPObservation) node, mt);
+            final SPNodeKey progKey = node.getProgram().getNodeKey();
+            final NodeValueCache<IP2Problems> nvc = (progKey == cacheKey) ? cache : NodeValueCache$.MODULE$.empty();
+
+            final scala.Tuple2<IP2Problems, NodeValueCache<IP2Problems>> tup;
+            tup = nvc.get(node, new scala.runtime.AbstractFunction1<ISPNode, IP2Problems>() {
```

Comment:
There is SAM support in 2.11 under `-Xfuture` I think, but we're still on 2.10 so I think no.